### ISSUE TITLE
Same repo but different additional dependencies

### ISSUE THREAD

**Issue by coldnight** [pre-commit/pre-commit#590]
I have two Git projects `A` and `B`, and those two use same [pre-commit repo](https://github.com/coldnight/pre-commit-pylint). But those two projects use different additional dependencies:

`.pre-commit-config.yaml` in `A`:

```yaml
-   repo: git@github.com:coldnight/pre-commit-pylint.git
    sha: 630e2662aabf3236fc62460b163d613c4bd1cfbc
    hooks:
    -   id: pylint-py3k
    -   id: pylint-score-limit
        args:
        - --limit=8.5
        - --rcfile=./.pylintrc
        additional_dependencies:
        - enum34; python_version<='3.4'
        - mock
```

`.pre-commit-config.yaml` in `B`:

```yaml
-   repo: git@github.com:coldnight/pre-commit-pylint.git
    sha: 630e2662aabf3236fc62460b163d613c4bd1cfbc
    hooks:
    -   id: pylint-py3k
    -   id: pylint-score-limit
        args:
        - --limit=8.5
        - --rcfile=./.pylintrc
        additional_dependencies:
        - enum34; python_version<='3.4'
        - requests
```

Here is my problem:

1. First I run `pre-commit` in project `A`, and the environment has installed
2. And then I run `pre-commit` in project `B`, and the environment has installed
3. When back project `A` and   run `pre-commit`, the installed environment has removed by above and need another installation(Huge slow!)

Any idea for this? Support different projects that use different home directories?

**Comment by asottile** [comment]
Ah yes this makes sense based on how the key for a repository is configured.

It currently creates one git repository for each `(repo, ref)` pair and then validates install state inside that cloned repository.  A workaround would be to use a *slightly* different repository name or ref.

There's some [recently added code](https://github.com/pre-commit/pre-commit/blob/7139a47c1ca968a2699e467279677fa77ad68aae/pre_commit/store.py#L133-L139) for local repositories that can probably be borrowed here such that different `additional_dependencies` repositories get their own directories.   Notice that here, the key becomes `('local:sorted,joined,deps', 'N/A')` -- perhaps something similar can be done for normal repositories as well!

An aside: pylint is a bit of a beast to get working with pre-commit as most of its checks aren't static analysis but *dynamic* analysis (it needs to be able to import your code to check it).  pylint often works better as a `local` / `system` / `script` hook which takes advantage of the currently activated virtualenv (and then has the ability to import and inspect your code directly).

**Comment by coldnight** [comment]
Thanks for your reply!

---

> An aside: pylint is a bit of a beast to get working with pre-commit as most of its checks aren't static analysis but dynamic analysis (it needs to be able to import your code to check it)

I agree with that.

---

>  pylint often works better as a local / system / script hook which takes advantage of the currently activated virtualenv (and then has the ability to import and inspect your code directly).

I will give my try.

**Comment by coldnight** [comment]
>  pylint often works better as a local / system / script hook which takes advantage of the currently activated virtualenv (and then has the ability to import and inspect your code directly).

It works. Thanks very much, and if there no plan for this, please let me know,  I will close this issue.

**Comment by asottile** [comment]
I do plan to fix this in a better way at some point so I'll keep this issue around -- thanks for pointing it out and I'm glad the workaround works well for you :)

**Comment by asottile** [comment]
Finally getting some time to fix this -- see #711

**Comment by coldnight** [comment]
Thanks for your awesome job!

**Comment by asottile** [comment]
This has been released as part of [v1.7.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.7.0)!
